src/main/java/org/thymeleaf/context/IJakartaWebContext.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * =============================================================================
+ *
+ *   Copyright (c) 2011-2018, The THYMELEAF team (http://www.thymeleaf.org)
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ * =============================================================================
+ */
+package org.thymeleaf.context;
+
+import jakarta.servlet.ServletContext;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import jakarta.servlet.http.HttpSession;
+
+/**
+ * <p>
+ *   Specialization of the {@link IContext} interface to be implemented by contexts used for template
+ *   processing in web environments.
+ * </p>
+ * <p>
+ *   Objects implementing this interface add to the usual {@link IContext} data the Servlet-API-related
+ *   artifacts needed to perform web-oriented functions such as URL rewriting or request/session access.
+ * </p>
+ * <p>
+ *   Note a class with this name existed since 1.0, but it was completely reimplemented
+ *   in Thymeleaf 3.0
+ * </p>
+ *
+ * @author Daniel Fern&aacute;ndez
+ *
+ * @since 3.0.0
+ *
+ */
+public interface IJakartaWebContext extends IContext {
```

Comment:
This approximation certainly does not break the current `IWebContext` interface contract, but I'm wondering how much of an issue that would be as I don't think any Thymeleaf extensions (esp. dialects) would directly use this interface — perhaps only in very specific extensions requiring access to low-level infrastructure such as the Spring Security one…

The first approximation I though for the duplication of this hierarchy was to remove all javax-related methods from `IWebContext` to a new `IJavaxWebContext` (or whichever name), leaving `IWebContext` as a mere marker of the fact that the context being evaluated was for a web application. This would make both `IJakartaWebContext` and `IJavaxWebContext` extend from `IWebContext` and could ease –just a theory though, you might have evaluated– ease the transition of all code where things like `instanceof IWebContext` are used for determining the _webapp-ness_ of the current execution (in the core, the `-springX` modules, and extension modules such as `thymeleaf-extras-springsecurityX`). I'm sure there will be a lot of code still requiring adjustment though, but perhaps this could make life easier?

Another possibility, which could be even less intrusive for code in Thymeleaf, would be to put an `org.thymeleaf.web.servlet*` inteface on top of the most commonly used javax/jakarta structures such as `HttpServletRequest`, `HttpSession`, etc, thus totally uncoupling most of Thymeleaf's code from these specific libraries, which would only be directly imported in the corresponding javax/jakarta implementations of these interfaces. Conceptually, I believe this would be my preferred approach because it would even open a nice window for _easy testing_ of web-related sections of Thymeleaf without the need to rely on mocking. Also, this would eliminate the need for duplicating this `IWebContext` hierarchy, replacing this need for the modification of this interface and all related subinterfaces and classes in order to use these new `org.thymeleaf.web.servlet.*` interfaces…

What do you think? Do you think this latest approach would be feasible?

Reply:
I don't see an advantage in terms of making life easier in breaking `IWebContext` to just be a marker of `IJavaxWebContext` and `IJakartaWebContext`. The reason is that there will be no common methods on `IWebContext` and so instanceof checks will be needed for both the javax and jakarta support.

This would be fixed if Thymeleaf abstracts out the HttpServletRequest and HttpSession, etc as you suggested. However, I don't know that this is a good idea. 

1) Jakarta started by coping the exact APIs of javax, but it will be adding new methods that will never be added to javax. If Thymeleaf tries to abstract out the methods it will just complicate things since Jakarta will have additional methods that are not supported by javax. 
2) Abstracting the APIs out will also add additional complexity to support multiple Jakarta versions as Jakarta adds new methods to their servlet APIs.
3) It would be a lot of work to extract out all the methods. Only extracting some of the methods would likely end up in additional feature requests.
4) The servlet APIs have been stable for years and it is only now that it has caused a problem.

Reply:
I'm going to give your suggestion a try. I'll post an update when I have it.